staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,339 @@
+	// to ensure that external requests are not served until the StorageVersion Informer and
+	// StorageVersionManager has synced
+	if !h.HasFinishedSync() {
+		handler.ServeHTTP(w, req)
+		return
+	}
+
+	gvr := schema.GroupVersionResource{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion, Resource: requestInfo.Resource}
+	if requestInfo.APIGroup == "" {
+		gvr.Group = "core"
+	}
+	serviceableByResp, err := h.findServiceableByServers(gvr, serverId)
+	if err != nil {
+		// this means that resource is an aggregated API or a CR since it wasn't found in SV informer cache, pass as it is
+		klog.Infof(fmt.Sprintf("no StorageVersion/APIServerID found for the GVR: %v skipping proxying", gvr))
+		handler.ServeHTTP(w, req)
+		return
+	}
+	// found the gvr locally, pass handler as it is
+	if serviceableByResp.locallyServiceable {
+		klog.V(4).Infof("resource can be served locally, skipping proxying")
+		handler.ServeHTTP(w, req)
+		return
+	}
+
+	gv := schema.GroupVersion{Group: gvr.Group, Version: gvr.Version}
+
+	// if no apiservers were found that could serve the request, pass as it to the next handler in the chain
+	// which should eventually return 404
+	if len(serviceableByResp.serviceableBy) == 0 {
+		klog.Errorf(fmt.Sprintf("GVR %v is not served by anything in this cluster", gvr))
+		responsewriters.ErrorNegotiated(apierrors.NewNotFound(schema.GroupResource{Group: gvr.Group, Resource: gvr.Resource},
+			fmt.Sprintf("%s.%s", gvr.Group, gvr.Resource)), s, gv, w, req)
+		return
+	}
+
+	// otherwise, randomly select an apiserver
+	rand := rand.Intn(len(serviceableByResp.serviceableBy))
+	apiserverId := serviceableByResp.serviceableBy[rand]
+
+	// finally proxy
+	ip, err := reconciler.GetEndpoint(apiserverId)
+	if err != nil {
+		klog.ErrorS(err, "error proxying request for the requested GVR")
+		responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable(fmt.Sprintf("Error proxying request for the requested GVR: %v, err: %v", gvr, err)), s, gv, w, req)
+	}
+
+	// TODO: currently only looking at the first element of ips, check how to handle other values
+	if ip == "" {
+		klog.Errorf("Failed to serve request: Found no endpoints in server leases for the remote apiserver")
+		responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Found no endpoints in servers leases for the remote apiserver"), s, gv, w, req)
+		return
+	}
+
+	if len(strings.Split(ip, ":")) != 2 {
+		klog.ErrorS(err, "error getting ip and port info of the remote server while proxying")
+		responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), s, gv, w, req)
+		return
+	}
+
+	h.proxyRequestToDestinationAPIServer(req, w, ip)
+
+}
+
+func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource, localAPIServerId string) (serviceableByResponse, error) {
+
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+	apiserversi, ok := h.svMap.Load(gvr)
```

Comment:
I found simpler to use a lock and a map,

Reply:
I realized I should have removed lock once I switched to sync map. Removed the lock now.